reactive/docker_registry.py

```diff
@@ -0,0 +1,221 @@
+
+    layer.docker_registry.stop_registry(name=name)
+    layer.docker_registry.configure_registry()
+    layer.docker_registry.start_registry()
+
+    report_active_status()
+
+
+@when('charm.docker-registry.configured')
+@when('endpoint.docker-registry.requests-pending')
+def handle_requests():
+    '''Set all the registry config that clients may care about.'''
+    registry = endpoint_from_flag('endpoint.docker-registry.requests-pending')
+    charm_config = hookenv.config()
+    port = charm_config.get('registry-port')
+    http_config = charm_config.get('http_config', '')
+
+    # tls config
+    url_prefix = 'http'
+    tls_enabled = False
+    tls_ca = None
+    if is_flag_set('charm.docker-registry.tls-enabled'):
+        url_prefix = 'https'
+        tls_enabled = True
+        cert_provider = endpoint_from_flag('cert-provider.ca.available')
+        if cert_provider:
+            tls_ca = cert_provider.root_ca_cert
+
+    # http config
+    if http_config and not http_config == "":
+        # When set, this config is our source of truth for the registry url
+        registry_url = http_config
+    else:
+        registry_url = '{}://{}:{}'.format(url_prefix, hookenv.unit_public_ip(), port)
+
+    for request in registry.requests:
+        hookenv.log('Sending {} and tls config to registry client.'.format(registry_url))
+        request.set_registry_config(registry_url=registry_url,
+                                    tls_enabled=tls_enabled,
+                                    tls_ca=tls_ca)
+    registry.mark_completed()
+
+
+@when('cert-provider.available')
+def request_certificates():
+    '''Request a new cert/key initally, and any time our SANs change.'''
+    cert_provider = endpoint_from_flag('cert-provider.available')
+
+    # Set the private ip of this unit as the Common Name for the cert.
+    # NB: Any 'http-host' config will be added to the SANs list; we
+    # want to ensure we always have a consistent CN regardless of config.
+    cert_cn = hookenv.unit_private_ip()
+
+    # Create a path safe name by removing path characters from the unit name.
+    cert_name = hookenv.local_unit().replace('/', '_')
+
+    # gather up all the alt names we want for our cert
+    proxy_ep = endpoint_from_flag('proxy.available')
+    sans = layer.docker_registry.get_tls_sans(proxy_ep.relation if proxy_ep else None)
+
+    # if our alt names have changed, request a new cert
+    if data_changed('tls_sans', sans):
+        hookenv.log('Requesting new cert for CN: {} with SANs: {})'.format(cert_cn, sans))
+        cert_provider.request_server_cert(cert_cn, sans, cert_name)
+
+
+@when('charm.docker-registry.configured')
+@when('cert-provider.server.certs.changed')
+def write_certs():
+    '''Write cert data to our configured location.'''
+    cert_provider = endpoint_from_flag('cert-provider.server.certs.changed')
+    ca = cert_provider.root_ca_cert
+    cert = cert_provider.server_certs[0]  # only requested one
```

Comment:
Is it possible to get the certificate is a non-0 position? https://github.com/juju-solutions/interface-tls-certificates/blob/master/requires.py#L174 What would happen if the same easyrsa is used by other deployments? Does this play any role?

Reply:
Would a dictionary work better for what the interface returns?

Reply:
Yup, we can use https://github.com/juju-solutions/interface-tls-certificates/blob/master/requires.py#L204 and look up the cert by CN.